### PR TITLE
Python frontend for style transfer

### PR DESCRIPTION
**Pull request by shreyajain17**
This PR does the following -
- Prints device name its training on (added in `style_transfer.cpp`)
- Pretrained models hosted on developer.apple.com are being pulled in the frontend and being passed to C++
- Added calculation `num_styles` also to python side
- Introduced a environment variable like for AC/OD for style_transfer
- Added the style_transfer_beta class

### CODE REVIEW

File: src/python/turicreate/toolkits/_pre_trained_models.py [hunk 1]
```diff
@@ -224,10 +224,41 @@ def __init__(self):
     def get_model_path(self):
         return self.model_path
 
+class StyleTransferTransformerCoreML():
+
+    def __init__(self):
+        self.name = 'transformer'
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/_pre_trained_models.py]: Underscore `_coreml`? Perhaps could be a problem.

File: src/python/turicreate/toolkits/_pre_trained_models.py [hunk 2]
```diff
@@ -224,10 +224,41 @@ def __init__(self):
...
+
+
+    def get_model_path(self):
+        return self.model_path
+
+
+class Vgg16CoreML():
+
+    def __init__(self):
+        self.name = 'vgg16'
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/_pre_trained_models.py]: Same ^

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 3]
```diff
@@ -634,6 +661,154 @@ def _raise_error_if_not_training_sframe(dataset, context_column):
...
         raise _ToolkitError("Context Image column must contain images")
 
+class StyleTransfer_beta(_Model):
+    """
+    A trained model using C++ implementation that is ready to use for classification or export to
+    CoreML.
+
+    This model should not be constructed directly.
+    """
+    _CPP_ACTIVITY_CLASSIFIER_VERSION = 1
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Change `_CPP_ACTIVITY_CLASSIFIER_VERSION` to `_CPP_STYLE_TRANSFER_VERSION `

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 4]
```diff
@@ -634,6 +661,154 @@ def _raise_error_if_not_training_sframe(dataset, context_column):
...
+
+    def __init__(self, model_proxy=None, name=None):
+        self.__proxy__ = model_proxy
+        self.__name__ = name
+
+    @classmethod
+    def _native_name(cls):
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: In custom model import we need a name here. Perhaps "style_transfer_beta" or the same name as above?

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 5]
```diff
@@ -634,6 +661,154 @@ def _raise_error_if_not_training_sframe(dataset, context_column):
...
+
+        Returns
+        -------
+        out : string
+            A description of the model.
+        """
+        return self.__class__.__name__
+
+    def _get_version(self):
+        return self._CPP_ACTIVITY_CLASSIFIER_VERSION
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Replace with `_CPP_STYLE_TRANSFER_VERSION `

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 6]
```diff
@@ -634,6 +661,154 @@ def _raise_error_if_not_training_sframe(dataset, context_column):
...
+
+        See Also
+        --------
+        save
+
+        Examples
+        --------
+        >>> model.export_coreml('StyleTransfer.mlmodel')
+        """
+        return self.__proxy__.export_to_coreml(filename, image_shape, include_flexible_shape)
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Make it as options as in c++.

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 7]
```diff
@@ -634,6 +661,154 @@ def _raise_error_if_not_training_sframe(dataset, context_column):
...
+        |   0    |   3   | Height: 256 Width: 256 |
+        |   1    |   0   | Height: 640 Width: 648 |
+        |   1    |   1   | Height: 640 Width: 648 |
+        |   1    |   2   | Height: 640 Width: 648 |
+        |   1    |   3   | Height: 640 Width: 648 |
+        +--------+-------+------------------------+
+        [8 rows x 3 columns]
+        """
+
+        return self.__proxy__.predict(images, style=None, verbose=True, max_size=800, batch_size = 4)
```
Comment (abhishekpratapa) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Follow the opts model defined in c++.

File: src/python/turicreate/toolkits/_pre_trained_models.py (hunk 1 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/_pre_trained_models.py]: Yes.

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py [hunk 8]
```diff
@@ -160,6 +163,11 @@ def create(style_dataset, content_dataset, style_feature=None,
     _raise_error_if_not_training_sframe(content_dataset, content_feature)
     _tkutl._handle_missing_values(style_dataset, style_feature, 'style_dataset')
     _tkutl._handle_missing_values(content_dataset, content_feature, 'content_dataset')
+
+    pretrained_resnet_model = _pre_trained_models.STYLE_TRANSFER_BASE_MODELS['resnet_mlmodel']()
+    pretrained_resnet_model_path = pretrained_resnet_model.get_model_path()
+    pretrained_vgg16_model = _pre_trained_models.STYLE_TRANSFER_BASE_MODELS['vgg_mlmodel']()
+    pretrained_vgg16_model_path = pretrained_vgg16_model.get_model_path()
```
Comment (nickjong) [inline src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Shouldn't we avoid downloading this if we're not using the CPP path?

File: src/python/turicreate/toolkits/style_transfer/style_transfer.py (hunk 4 above)
Comment (nickjong) [reply src/python/turicreate/toolkits/style_transfer/style_transfer.py]: Please consult with what @fareeha did for OD and AC to make this work for both settings of the feature flag

File: src/python/turicreate/toolkits/_pre_trained_models.py [hunk 9]
```diff
@@ -199,30 +199,40 @@ class StyleTransferTransformer():
...
+        self.source_md5 = { 'mxnet' : 'ac232afa6d0ead93a8c75b6c455f6dd3',
+                            'coreml': 'e0f3adaa9952ecc7d96f5e4eefb0d690' }
 
 
-    def get_model_path(self):
-        return self.model_path
+    def get_model_path(self, format):
+        assert(format in ('coreml', 'mxnet'))
+        if (format in 'coreml') :
+            url =  _urlparse.urljoin(MODELS_URL_ROOT, 'transformer.mlmodel')
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_pre_trained_models.py]: Should I worry that the base names are different between CoreML and MXNet? Is this the only transformer network we'll ever publish here?

File: src/python/turicreate/toolkits/_pre_trained_models.py [hunk 10]
```diff
@@ -199,30 +199,40 @@ class StyleTransferTransformer():
...
 
 
-    def get_model_path(self):
-        return self.model_path
+    def get_model_path(self, format):
+        assert(format in ('coreml', 'mxnet'))
+        if (format in 'coreml') :
+            url = _urlparse.urljoin(MODELS_URL_ROOT, 'vgg16.mlmodel')
+        else:
+            url = _urlparse.urljoin(MODELS_URL_ROOT, 'vgg16-conv1_1-4_3.params')
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_pre_trained_models.py]: What's the significance of `-conv1_1-4_3` in the MXNet case?

Comment (abhishekpratapa) [reply src/python/turicreate/toolkits/_pre_trained_models.py]: It's specifying the subset of layers we use in VGG16. We only use the first 12 or so layers of the VGG16 network. May be worth renaming.

**Pull request merged**